### PR TITLE
Add `_from_column_like_self` factory

### PR DESCRIPTION
**Pull request by isVoid**
Follow up to #9558

On a return trip from libcudf, it is a common pattern for cudf frame to apply its own metadata to the columns. This PR generalizes this procedure as a new factory function `_from_colums_like_self`

### CODE REVIEW

File: python/cudf/cudf/core/frame.py [hunk 1]
```diff
@@ -163,6 +163,16 @@ def _from_columns(
 
         return cls._from_data(data, index)
 
+    def _from_columns_like_self(
+        self,
+        columns: List[ColumnBase],
+        column_names: List[str],
+        index_names: Optional[List[str]] = None,
+    ):
```
Comment (rgsl888prabhu) [inline python/cudf/cudf/core/frame.py:171]: A small description would be better.

**Review by rgsl888prabhu — APPROVED** [review]
A small suggestion

File: python/cudf/cudf/core/frame.py (hunk 1 above)
Comment (vyasr) [reply python/cudf/cudf/core/frame.py:171]: Agreed, this will become common enough to need a docstring.

File: python/cudf/cudf/core/frame.py [hunk 2]
```diff
@@ -163,6 +163,16 @@ def _from_columns(
...
+        col = self.__class__._from_columns(columns, column_names, index_names)
+        col._copy_type_metadata(self, include_index=bool(index_names))
+        return col
```
Comment (vyasr) [inline python/cudf/cudf/core/frame.py:174]:
```suggestion
        frame = self.__class__._from_columns(columns, column_names, index_names)
        return frame._copy_type_metadata(self, include_index=bool(index_names))
```

File: python/cudf/cudf/core/frame.py [hunk 3]
```diff
@@ -163,6 +163,16 @@ def _from_columns(
 
         return cls._from_data(data, index)
 
+    def _from_columns_like_self(
+        self,
+        columns: List[ColumnBase],
+        column_names: List[str],
+        index_names: Optional[List[str]] = None,
```
Comment (vyasr) [inline python/cudf/cudf/core/frame.py:170]: Is the plan to eventually specialize this in `IndexedFrame`?

File: python/cudf/cudf/core/_base_index.py [hunk 4]
```diff
@@ -1444,7 +1444,8 @@ def drop_duplicates(
         """
 
         # This utilizes the fact that all `Index` is also a `Frame`.
-        result = self.__class__._from_columns(
+        # Except RangeIndex.
```
Comment (vyasr) [inline python/cudf/cudf/core/_base_index.py:1447]: Yeah, we'll probably want to move this impl eventually. No specific urgency though.

**Review by vyasr — CHANGES_REQUESTED** [review]
I like keeping this PR's scope narrow for now since it's targeting 22.02. We'll need to make a follow-up PR that looks through all of the classes inheriting from `Frame` and substitutes this new factory everywhere that it could be used. I think there are a lot more places. Additionally, we'll probably find some places where `_from_columns` is being called without a subsequent call to `_apply_type_metadata`, so we should be able to fix some bugs by doing that scan.

File: python/cudf/cudf/core/frame.py (hunk 3 above)
Comment (isVoid) [reply python/cudf/cudf/core/frame.py:170]: I think we can, and we need to provide support to Index/MultiIndex too, in that case we would need create its counterpart in `_base_index`.

Comment (isVoid) [reply python/cudf/cudf/core/frame.py:170]: We would need to move `_from_columns` too.

File: python/cudf/cudf/core/_base_index.py (hunk 4 above)
Comment (isVoid) [reply python/cudf/cudf/core/_base_index.py:1447]: Let's introduce this in this round, and move it in the next cycle.

**Review by rgsl888prabhu — APPROVED** [review]

**Review by trxcllnt — APPROVED** [review]

**Review by vyasr — APPROVED** [review]

**Comment by isVoid** [comment]
rerun tests

**Comment by vyasr** [comment]
rerun tests

**Comment by galipremsagar** [comment]
rerun tests

**Comment by isVoid** [comment]
rerun tests

**Comment by shwina** [comment]
@gpucibot merge

**Comment by galipremsagar** [comment]
@gpucibot merge

**Pull request merged**